consensus/istanbul/uptime/uptime.go

```diff
@@ -0,0 +1,242 @@
+// First blocks of an epoch need to be skipped since we can't assess the last `lookbackWindow` block for validators
+// as those are froma different epoch.
+// Similarly, last block of epoch is skipped since we can't obtaine the signer for it; as they are in the next block
+func GetUptimeMonitoringWindow(epochNumber uint64, epochSize uint64, lookbackWindowSize uint64) (uint64, uint64) {
+	if epochNumber == 0 {
+		panic("no monitoring window for epoch 0")
+	}
+
+	epochFirstBlock, _ := istanbul.GetEpochFirstBlockNumber(epochNumber, epochSize)
+	epochLastBlock := istanbul.GetEpochLastBlockNumber(epochNumber, epochSize)
+
+	// first block to monitor:
+	// We need to wait for the completion of the first window with the start window's block being the
+	// 2nd block of the epoch, before we start tallying the validator score for epoch "epochNumber".
+	// We can't include the epoch's first block since it's aggregated parent seals
+	// is for the previous epoch's valset.
+	firstBlockToMonitor := epochFirstBlock + 1 + (lookbackWindowSize - 1)
+
+	// last block to monitor:
+	// We stop tallying for epoch "epochNumber" at the second to last block of that epoch.
+	// We can't include that epoch's last block as part of the tally because the epoch val score is calculated
+	// using a tally that is updated AFTER a block is finalized.
+	// Note that it's possible to count up to the last block of the epoch, but it's much harder to implement
+	// than couting up to the second to last one.
+	lastBlockToMonitor := epochLastBlock - 1
+
+	return firstBlockToMonitor, lastBlockToMonitor
+}
+
+// Monitor is responsible for monitoring uptime
+// by processing blocks
+type Monitor struct {
+	epochSize      uint64
+	lookbackWindow uint64
+
+	logger log.Logger
+	store  Store
+}
+
+// NewMonitor creates a new uptime monitor
+func NewMonitor(store Store, epochSize, lookbackWindow uint64) *Monitor {
+	return &Monitor{
+		epochSize:      epochSize,
+		lookbackWindow: lookbackWindow,
+		store:          store,
+		logger:         log.New("module", "uptime-monitor"),
+	}
+}
+
+// MonitoringWindow returns the monitoring window for the given epoch in the format
+// [firstBlock, lastBlock] both inclusive
+func (um *Monitor) MonitoringWindow(epoch uint64) (uint64, uint64) {
+	return GetUptimeMonitoringWindow(epoch, um.epochSize, um.lookbackWindow)
+}
+
+// MonitoringWindowSize returns the size (in blocks) of the monitoring window
+func (um *Monitor) MonitoringWindowSize(epoch uint64) uint64 {
+	start, end := um.MonitoringWindow(epoch)
+	return end - start + 1
+}
+
+// ComputeValidatorsUptime retrieves the uptime score for each validator for a given epoch
+func (um *Monitor) ComputeValidatorsUptime(epoch uint64, valSetSize int) ([]*big.Int, error) {
+	logger := um.logger.New("func", "Backend.updateValidatorScores", "epoch", epoch)
+	logger.Trace("Updating validator scores")
+
+	// The denominator is the (last block - first block + 1) of the val score tally window
```

Comment:
Might be clearer if we explicitly say this is the maximum possible uptime score

Reply:
agree that `denominator` doesn't really inform anything.

It's not the max uptime score (since score it's a ratio), but the total blocks that we are monitoring

So, what about:
```
totalMonitoredBlocks